File: 28.py
def longestWPI(hours):
    prefix = 0
    seen = {}
    res = 0

    for i, h in enumerate(hours):
        prefix += 1 if h > 8 else -1
        if prefix > 0:
            res = i + 1
        elif (prefix - 1) in seen:
            res = max(res, i - seen[prefix - 1])
        if prefix not in seen:
            seen[prefix] = i
    return res

# Example
print(longestWPI([9,9,6,0,6,6,9]))  # Output: 3

File: 28.js
function longestWPI(hours) {
  let prefix = 0;
  const seen = new Map();
  let res = 0;

  for (let i = 0; i < hours.length; i++) {
    prefix += hours[i] > 8 ? 1 : -1;
    if (prefix > 0) {
      res = i + 1;
    } else if (seen.has(prefix - 1)) {
      res = Math.max(res, i - seen.get(prefix - 1));
    }
    if (!seen.has(prefix)) seen.set(prefix, i);
  }

  return res;
}

console.log(longestWPI([9, 9, 6, 0, 6, 6, 9])); // Output: 3
